File: Array/AlgoExpert/14. Apartment Hunting.py
def apartmentHunting(blocks, reqs):
    minDistanceFromBlocks = list(map(lambda req: getMinDistances(blocks,req)),  reqs)
    maxDistanceAtBlocks = getMaxDistancesAtBlocks(blocks, minDistanceFromBlocks)
    return getIdxAtMinValue(maxDistanceAtBlocks)



def getMinDistances(blocks, req):
    minDistances = [0 for block in blocks]
    closestReqIdx = float("inf")
    for i in range(len(blocks)):
        if blocks[i][req]:
            closestReqIdx = i
        minDistances[i] = distanceBetween(i, closestReqIdx)
    
    for i in reversed(range(len(blocks))):
        if blocks[i][req]:
            closestReqIdx = i
        minDistances[i] = min(minDistances[i],distanceBetween(i, closestReqIdx))
    return minDistances

def getMaxDistancesAtBlocks(blocks, minDistanceFromBlocks):
    maxDistancesAtBlocks = [0 for block in blocks]
    for i in range(len(blocks)):
        minDistancesAtBlocks = list(map(lambda distances: distances[i], minDistanceFromBlocks))
        maxDistancesAtBlocks[i] = max(minDistancesAtBlocks)
    return maxDistancesAtBlocks

def getIdxAtMinValue(array):
    idxAtMinValue = 0
    minValue = float("inf")
    for i in range(len(array)):
        currentValue = array[i]
        if currentValue < minValue:
            minValue = currentValue
            idxAtMinValue = i
    return idxAtMinValue

def distanceBetween(a,b):
    return abs(a-b)

File: Array/AlgoExpert/14. Apartment Hunting.cpp
#include <vector>
#include <unordered_map>

using namespace std;

int distancebetween(int a, int b)
{
    return abs(a-b);
}
int minValueIndex(vector<int> A)
{
    int minimum = INT_MAX;
    int minIndex = 0;
    int n = A.size();

    for(int i=0;i<n;i++)
    {
        if(minimum > A[i])
        {
            minimum = A[i];
            minIndex = i;
        }
    }

    return minIndex;
}
int apartmentHunting(vector<unordered_map<string, bool>> blocks,
                     vector<string> reqs) 
{
	vector<int> MaxdistanceForEveryBlock(blocks.size(),INT_MIN);

    for(int i=0; i < blocks.size(); i++)
    {
        for(auto req: reqs)
        {
            int closestRequiredDistance = INT_MAX;

            for(int j=0; j < blocks.size(); j++)
            {
                if(blocks[j][req])
                {
                    closestRequiredDistance = min(closestRequiredDistance,distancebetween(i,j));
                }
            }

            MaxdistanceForEveryBlock[i] = max(MaxdistanceForEveryBlock[i],closestRequiredDistance);
        }
    }

    return minValueIndex(MaxdistanceForEveryBlock);
}